src/codec.rs

```diff
@@ -885,9 +933,118 @@ impl<T: Encode, const N: usize> Encode for [T; N] {
 }
 
 impl<T: Decode, const N: usize> Decode for [T; N] {
-	#[inline]
+	#[inline(always)]
 	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
-		decode_array(input)
+		let mut array = MaybeUninit::uninit();
+		Self::decode_into(DecodeContext::new(), input, &mut array)?;
+
+		// SAFETY: `decode_into` succeeded, so the array is initialized.
+		unsafe {
+			Ok(array.assume_init())
+		}
+	}
+
+    fn decode_into<I: Input>(ctx: DecodeContext, input: &mut I, dst: &mut MaybeUninit<Self>) -> Result<DecodeFinished, Error> {
+		let is_primitive = match <T as Decode>::TYPE_INFO {
+			| TypeInfo::U8
```

Comment:
This trick cannot be used for booleans because they have a niche. That is, not every bit pattern for booleans is valid: only `0b00000001` and `0b00000000` are, so we can't just read them in bulk.

(Well, we *can*, but we'd need to add an extra loop that would verify that every `byte & !1 == 0`, and with SIMD such a loop would most likely be pretty cheap. Probably not worth the effort, and out of scope of this PR. :P)